### PR TITLE
[DiffusionPipeline] Deprecate not throwing error when loading non-existant variant

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

Thanks for pointing out this problem @vladmandic . Upon thinking about it, I don't think we should even allow default to fp32 when trying to load "fp16" variants. Added a deprecation warning now.

Note that one can very easily test with this function if the variant is present: https://github.com/huggingface/diffusers/pull/4011/files#r1257526906

So ideally this should be done before downloading - e.g. via:

```py
from huggingface_hub import model_info

info = model_info("<your/repo/id>")
filenames = {sibling.rfilename for sibling in info.siblings}
```

We can however also very easily check post-download what variants have been downloaded as shown here: https://github.com/huggingface/diffusers/pull/4011#discussion_r1257526906

### CODE REVIEW

File: tests/pipelines/test_pipelines.py [hunk 1]
```diff
@@ -1361,6 +1363,29 @@ def test_optional_components(self):
...
+        assert variant in str(warning_context.warning)
+
+        def get_all_filenames(directory):
+            filenames = glob.glob(directory + "/**", recursive=True)
+            filenames = [f for f in filenames if os.path.isfile(f)]
+            return filenames
+
+        filenames = get_all_filenames(str(cached_folder))
+
+        all_model_files, variant_model_files = variant_compatible_siblings(filenames, variant=variant)
```
Comment (patrickvonplaten) [inline tests/pipelines/test_pipelines.py:1383]: That's how one can easily test which variant was downloaded

**Review by sayakpaul — APPROVED** [review]

File: tests/pipelines/test_pipelines.py [hunk 2]
```diff
@@ -1361,6 +1363,29 @@ def test_optional_components(self):
...
+
+        # make sure that none of the model names are variable model names
```
Comment (sayakpaul) [inline tests/pipelines/test_pipelines.py:1385]:
```suggestion
        # make sure that none of the model names are variant model names
```

**Comment by vladmandic** [comment]
LGTM, thanks @patrickvonplaten

File: src/diffusers/pipelines/pipeline_utils.py [hunk 3]
```diff
@@ -1213,6 +1213,15 @@ def download(cls, pretrained_model_name, **kwargs) -> Union[str, os.PathLike]:
             filenames = {sibling.rfilename for sibling in info.siblings}
             model_filenames, variant_filenames = variant_compatible_siblings(filenames, variant=variant)
 
+            if len(variant_filenames) == 0 and variant is not None:
+                deprecation_message = (
+                    f"You are trying to load the model files of the `variant={variant}`, but no such modeling files are available."
+                    f"The default model files: {model_filenames} will be loaded instead. Make sure to not load from `variant={variant}`"
+                    "if such variant modeling files are not available. Doing so will lead to an error in v0.22.0 as defaulting to non-variant"
+                    "modeling files will is deprecated."
```
Comment (pcuenca) [inline src/diffusers/pipelines/pipeline_utils.py:1221]:
```suggestion
                    "modeling files is deprecated."
```

**Review by pcuenca — APPROVED** [review]

**Pull request merged**